### ISSUE TITLE
include the function provided  in https://github.com/dfeich/org-babel-examples/blob/master/library-of-babel/dfeich-lob.org into the pg

### ISSUE THREAD

**Issue by ouboub** [dfeich/org-listcruncher#9]
Hi
First of all a big thank you for this code. I wish I had found it earlier!

The example as found in the documentation such as 
```
#+BEGIN_SRC elisp :results output raw :var listname="lstTest" :exports both 
     (princ (orgtbl-to-orgtbl
     	(append (org-listcruncher-to-table listname) '(hline ("Total")))
     	'(:tend "#+TBLFM: @>$3=vsum(@I..@II)::@>$2=vsum(@I..@II)")))
#+END_SRC
```

work nicely, the example

```
#+BEGIN_SRC elisp :results output raw :var listname="lstTest" :exports both :post lobPostAlignTables(*this*)
     (princ (orgtbl-to-orgtbl
     	(append (org-listcruncher-to-table listname) '(hline ("Total")))
     	'(:tend "#+TBLFM: @>$3=vsum(@I..@II)::@>$2=vsum(@I..@II)")))
#+END_SRC

```

even better. However for that I would need the code examples of 
https://github.com/dfeich/org-babel-examples/blob/master/library-of-babel/dfeich-lob.org

is there any change that you include that functions as lisp functions in your package?

regards

Uwe Brauer

**Comment by dfeich** [comment]
Dear Uwe,

sorry for only replying now. I'm happy that you find the code useful. I'm also still using this continuously for my planning and recently we even did planning in a group where we worked on such a list with ~10 people.

I think that the function `lobPostAlignTables` does not really fit thematically into the package, also it would acquire an even more unwieldy name. Actually this kind of function is so commonly useful that I had asked myself why is it not provided by org directly, since there are so many occasions where one wants to produce a table as a result, and then have org evaluate and iterate it.

Would it be ok for you to just define the function somewhere inside of your document in a separate code block? Else having these kind of functions available in a central library of babel is not such a bad solution. It's org's poor man solution for providing a library mechanism... a bit unsatisfactory but that's what it supports at the moment.

I may still consider adding the function, maybe. Maybe it makes the package more easily usable, and it's a small function only and will not so much add to the normal code bloat.

What do you think?

Derek

**Comment by ouboub** [comment]
>>> "DF" == Derek Feichtinger ***@***.***> writes:

Hello Derek,


> Dear Uwe,
> sorry for only replying now. I'm happy that you find the code useful.
> I'm also still using this continuously for my planning and recently we
> even did planning in a group where we worked on such a list with ~10
> people.


That is precisely what I have to do in the coming days, and that is why
I am really glad to have found your pkg, thanks

> I think that the function `lobPostAlignTables` does not really fit
> thematically into the package, also it would acquire an even more
> unwieldy name. Actually this kind of function is so commonly useful
> that I had asked myself why is it not provided by org directly, since
> there are so many occasions where one wants to produce a table as a
> result, and then have org evaluate and iterate it.

> Would it be ok for you to just define the function somewhere inside of
> your document in a separate code block? Else having these kind of
> functions available in a central library of babel is not such a bad
> solution. It's org's poor man solution for providing a library
> mechanism... a bit unsatisfactory but that's what it supports at the
> moment.


The point of having it in the document is that then org asks me two
times to evaluate a function. 

Now I have a confession to make, I tried to set it up as a common lisp
function in my private library but I failed.

It seem that the function requires a parameter/variable, since it is called as 

#+BEGIN_SRC elisp :results output raw :var listname="baremo" :exports both :post lobPostAlignTables(*this*)

lobPostAlignTables(*this*) Not entirely sure what that *this* is
supposed to be, if I replace it by anything else the code does not work
and if replace it by () I obtain the entry 
#+header: :var text="|5|22222|\n|0||\n|12|45|\n|---\n|||\n#+TBLFM:@***@***.***@-1)\n\n|1|22222|\n|0||\n|12|45|\n"

So a native attempt of defining 

(defun lobPostAlignTables ()
     (with-temp-buffer
       (erase-buffer)
       (cl-assert text nil "PostAlignTables received nil instead of text ")
       (insert text)
       (beginning-of-buffer)
       (org-mode)
       (while
           (search-forward-regexp org-table-any-line-regexp nil t)
         (org-table-align)
         (org-table-recalculate 'iterate)
         (goto-char (org-table-end)))
       (buffer-string))

Does not make much sense, it should be 

(defun lobPostAlignTables (table)
     (with-temp-buffer
       (erase-buffer)
       (cl-assert text nil "PostAlignTables received nil instead of text ")
       (insert text)
       (beginning-of-buffer)
       (org-mode)
       (while
           (search-forward-regexp org-table-any-line-regexp nil t)
         (org-table-align)
         (org-table-recalculate 'iterate)
         (goto-char (org-table-end)))
       (buffer-string))

But there is no variable in the code. So I am puzzled.

regards

Uwe

**Comment by ouboub** [comment]
Hello Derek

The following kills me (but I don't understand with-temp-buffer well
enough)
Works

(defun test-var (text)
(insert text))

(test-var "hello")


Does not work

(defun lobPostAlignTables (text)
     (with-temp-buffer
       (erase-buffer)
       (cl-assert text nil "PostAlignTables received nil instead of text ")
       (insert text)
       (beginning-of-buffer)
       (org-mode)
       (while
           (search-forward-regexp org-table-any-line-regexp nil t)
         (org-table-align)
         (org-table-recalculate 'iterate)
         (goto-char (org-table-end)))
       (buffer-string))


Uwe

**Comment by dfeich** [comment]
Hi Uwe,

I just prepared a ready made org document with explanations for you. You should be able to run it directly. The problem is just that the "post" is calling the code using the conventions for calling another babel block (like for #+CALL).  I had to zip the file because github does not allow org attachments.

Hope that helps. Else you can also contact me directly using the mail address given in the file header.
[for-uwe.org.gz](https://github.com/dfeich/org-listcruncher/files/6695840/for-uwe.org.gz)

Best regards,
Derek

**Comment by dfeich** [comment]
Hello Uwe,

some more comments to your questions

> Hello Derek The following kills me (but I don't understand with-temp-buffer well enough)

Your code will work, but maybe not as you intend it:

```elisp
 (defun test-var (text) (insert text))

 (test-var "hello")
```
That actually should work. Do it in the *scratch* buffer, do an `C-x C-e` after the defun line to define the function. When you execute your test call of the function below, it will just insert the hello into your current buffer at the cursor position... which will be the scratch buffer.

The `with-temp-buffer` just creates a temporary text buffer in emacs. That is the way one usually does most of the string manipulations (except the most primitive ones), because emacs has a lot of good functions that work around the concept of buffers. Usually you just `insert` the text into the temp buffer, do whatever you need, and then return the finished text as the return value. That is done by `buffer-string` in the current context... it just takes the whole contents of the buffer and returns them. Emacs will take care of then disposing of the temp buffer after its form is finished.

I added to the document also a `Local Variable` setting at the end that will configure an emacs setting so that you will not get asked for confirming the evaluation when working within the document. This is what I usually do for this kind of work.

Best regards,
Derek

**Comment by dfeich** [comment]
Hello Uwe,

I had some time today and implemented a way that can do the formula processing inside of list cruncher. I just pushed the commit mentioned within this thread (above) and released version 1.4. I will close the issue, now.

Best regards, Derek

**Comment by ouboub** [comment]
Great, thanks, I will pull and have a look. If I find problems, I will let you know. 
Thanks very much for implementing it.
Regards
Uwe
